mitiq/folding_cirq.py

```diff
@@ -339,26 +339,51 @@ def fold_method(circuit: Circuit, stretch: float, **kwargs):
 
 
 # Circuit level folding
-def unitary_folding(circuit: Circuit, stretch: float) -> Circuit:
-    """Applies global unitary folding and a final partial folding of the input circuit.
-    Returns a circuit of depth approximately equal to stretch*len(circuit).
-    The stretch factor can be any real number >= 1."""
+def fold_global(circuit: Circuit,
+                stretch: float,
+                finish_fold_method: Callable[[Circuit, float, Tuple[Any]], Circuit] = fold_gates_from_left,
+                finish_fold_method_args: Tuple[Any] = ()
+                ) -> Circuit:
+    """Returns a folded circuit obtained by folding the global unitary of the input circuit.
 
-    if not (stretch >= 1):
-        raise ValueError("The stretch factor must be a real number >= 1.")
+    A single unitary fold corresponds to a stretch of 3. If stretch (mod 3) > 1, a local folding method is called
+    until the desired stretch factor is reached.
 
-    # determine the number of integer foldings and the final fractional_stretch
-    num_foldings, fractional_stretch = divmod(stretch - 1, 2)
+    The returned folded circuit has a number of gates approximately equal to stretch * len(circuit).
 
-    # integer circuit folding
-    eye = Circuit()
-    for _ in range(int(num_foldings)):
-        eye += inverse(circuit) + circuit
 
-    # partial circuit folding.
-    depth = len(circuit)
-    fractional_depth = int(depth * fractional_stretch / 2)
-    if fractional_depth != 0:
-        eye += inverse(circuit[-fractional_depth:]) + circuit[-fractional_depth:]
+    Parameters
+    ----------
+        circuit: Circuit to fold.
+        stretch: Factor to stretch the circuit by.
+        finish_fold_method: A local folding method which is called to fold remaining
+                            gates to reach the input stretch factor, if necessary.
+        finish_fold_method_args: Any arguments to pass into finish_fold_method.
+    """
+
+    if not (stretch >= 1):
+        raise ValueError("The stretch factor must be a real number >= 1.")
 
-    return circuit + eye
+    if not circuit.are_all_measurements_terminal():
+        raise ValueError(f"Input circuit contains intermediate measurements and cannot be folded.")
+
+    folded = deepcopy(circuit)
+    measurements = _pop_measurements(folded)
+    base_circuit = deepcopy(folded)
+
+    # Determine the number of global folds and the final fractional stretch
+    num_global_folds = int(stretch // 3)
+    fractional_stretch = stretch % 3
```

Comment:
I think there is a problem here.  The integer stretch factors are = 1 + 2 n .  So they can be any odd integer number and not just multiples of 3.   So I think we need the previous 
`num_foldings, fractional_stretch = divmod(stretch - 1, 2)` or something equivalent to this.

Reply:
Fixed.